shiny/_template_utils.py

```diff
@@ -104,34 +103,173 @@ def download_and_extract_zip(url: str, temp_dir: Path):
     with zipfile.ZipFile(zip_file_path, "r") as zip_file:
         zip_file.extractall(temp_dir)
 
+    items = list(temp_dir.iterdir())
+
+    # If we unzipped a single directory, return the path to that directory.
+    # This avoids much nonsense in trying to guess the directory name, which technically
+    # can be derived from the zip file URL, but it's not worth the effort.
+    directories = [d for d in items if d.is_dir()]
+    files = [f for f in items if f.is_file() and f.name != "repo.zip"]
+
+    # We have exactly one directory and no other files
+    if len(directories) == 1 and len(files) == 0:
+        return directories[0]
+
+    return temp_dir
+
 
 def use_git_template(
-    url: str, mode: Optional[str] = None, dest_dir: Optional[Path] = None
+    url: str,
+    template: str | None = None,
+    mode: str | None = None,
+    dest_dir: Path | None = None,
 ):
     # Github requires that we download the whole repository, so we need to
     # download and unzip the repo, then navigate to the subdirectory.
 
-    parsed_url = urlparse(url)
-    path_parts = parsed_url.path.strip("/").split("/")
-    repo_owner, repo_name, _, branch_name = path_parts[:4]
-    subdirectory = "/".join(path_parts[4:])
+    spec = parse_github_arg(url)
+    spec_cli = f"{spec.repo_owner}/{spec.repo_name}"
+    if spec.ref != "HEAD":
+        spec_cli = f"{spec_cli}@{spec.ref}"
+    if spec.path:
+        spec_cli = f"{spec_cli}:{spec.path}"
 
-    zip_url = f"https://github.com/{repo_owner}/{repo_name}/archive/refs/heads/{branch_name}.zip"
+    click.echo(cli_info(f"Using GitHub repository {cli_field(spec_cli)}."))
 
     with tempfile.TemporaryDirectory() as temp_dir:
         temp_dir = Path(temp_dir)
-        download_and_extract_zip(zip_url, temp_dir)
+        template_dir = temp_dir
+        success = False
+
+        # Github uses different formats on the `archive/` endpoints for
+        # refs/heads/{branch}.zip, refs/tags/{tag}.zip, or just {commit}.zip.
+        # It's hard to tell in advance which one is intended, so we try all three in
+        # succession, using the first that works.
+        for zip_url in github_zip_url(spec):
+            try:
+                template_dir = download_and_extract_zip(zip_url, temp_dir)
+                success = True
+                break
+            except Exception:
+                pass
```

Comment:
Could accumulate execptions and include them if not success?

Reply:
```
❯ shiny create --github baslkjsadfas/aslalskdasdf
ℹ Using GitHub repository baslkjsadfas/aslalskdasdf.
Error: Failed to download repository from GitHub baslkjsadfas/aslalskdasdf. Please check the URL or GitHub spec and try again.
We tried the following URLs
  https://github.com/baslkjsadfas/aslalskdasdf/archive/refs/heads/HEAD.zip:
    HTTP Error 404: Not Found for url: https://github.com/baslkjsadfas/aslalskdasdf/archive/refs/heads/HEAD.zip
  https://github.com/baslkjsadfas/aslalskdasdf/archive/refs/tags/HEAD.zip:
    HTTP Error 404: Not Found for url: https://github.com/baslkjsadfas/aslalskdasdf/archive/refs/tags/HEAD.zip
  https://github.com/baslkjsadfas/aslalskdasdf/archive/HEAD.zip:
    HTTP Error 404: Not Found for url: https://github.com/baslkjsadfas/aslalskdasdf/archive/HEAD.zip
```